pkg/agent/packetcapture/packetcapture_controller_linux.go

```
@@ -0,0 +1,884 @@
+	options := CaptureOptions{
+		MaxCaptureLength: maxSnapshotBytes,
+		Promiscuous:      true,
+	}
+	packets, err := source.Capture(&options)
+	if err != nil {
+		klog.ErrorS(err, "failed to start capture")
+		return err
+	}
+
+	timer := time.NewTicker(timeout)
+	defer timer.Stop()
+
+	for {
+		select {
+		case packet := <-packets:
+			if captureState.capturedPacketsNum == captureState.targetCapturedPacketsNum {
+				return nil
+			}
+			captureState.capturedPacketsNum++
+			ci := gopacket.CaptureInfo{
+				Timestamp:     time.Now(),
+				CaptureLength: len(packet.Data()),
+				Length:        len(packet.Data()),
+			}
+			err = captureState.pcapngWriter.WritePacket(ci, packet.Data())
+			if err != nil {
+				return fmt.Errorf("couldn't write packet: %w", err)
+			}
+			klog.V(5).InfoS("capture packets", "name", captureState.name, "count",
+				captureState.capturedPacketsNum, "len", ci.Length)
+
+			reachTarget := captureState.capturedPacketsNum == captureState.targetCapturedPacketsNum
+			// use rate limiter to reduce the times we need to update status.
+			if reachTarget || captureState.updateRateLimiter.Allow() {
+				pc, err := c.packetCaptureLister.Get(captureState.name)
+				if err != nil {
+					return fmt.Errorf("get PacketCapture failed: %w", err)
+				}
+				// if reach the target. flush the file and upload it.
+				if reachTarget {
+					path := os.Getenv("POD_NAME") + ":" + nameToPath(pc.Name)
+					if err = captureState.pcapngWriter.Flush(); err != nil {
+						return err
+					}
+					if pc.Spec.FileServer != nil {
+						err = c.uploadPackets(pc, captureState.pcapngFile)
+						klog.V(4).InfoS("Upload captured packets", "name", pc.Name, "path", path)
+
+					}
+					// update capture result.
+					if updateErr := c.updatePacketCaptureStatus(pc.Name, captureState.capturedPacketsNum, path, err); updateErr != nil {
+						return updateErr
+					}
+					if err != nil {
+						return err
+					}
+					if err := captureState.pcapngFile.Close(); err != nil {
```

I feel we can just log the failure instead of a returned error.  A non-closed file should still be deleted successfully later.

---

updated